### PR TITLE
Added support for HPU device stats monitor

### PR DESCRIPTION
**Pull request by jerome-habana**
Signed-off-by: Jerome <janand@habana.ai>

## What does this PR do?

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source-pytorch/accelerators/hpu_intermediate.rst [hunk 1]
```diff
@@ -66,3 +66,32 @@ This enables advanced users to provide their own BF16 and FP32 operator list ins
...
+
+The below snippet shows how DeviceStatsMonitor can be enabled.
+
+.. code-block:: python
+
+    from pytorch_lightning import Trainer
+    from pytorch_lightning.callbacks import DeviceStatsMonitor
+
+    device_stats = DeviceStatsMonitor()
+    trainer = Trainer(callbacks=[device_stats])
```
Comment (kaushikb11) [inline docs/source-pytorch/accelerators/hpu_intermediate.rst:97]:
```suggestion
    trainer = Trainer(accelerator="hpu", callbacks=[device_stats])
```

File: src/pytorch_lightning/accelerators/hpu.py [hunk 2]
```diff
@@ -39,9 +39,24 @@ def setup_environment(self, root_device: torch.device) -> None:
             raise MisconfigurationException(f"Device should be HPU, got {root_device} instead.")
 
     def get_device_stats(self, device: Union[str, torch.device]) -> Dict[str, Any]:
```
Comment (kaushikb11) [inline src/pytorch_lightning/accelerators/hpu.py:41]: Could we add reference to an API doc here?

File: docs/source-pytorch/accelerators/hpu_intermediate.rst [hunk 3]
```diff
@@ -66,3 +66,32 @@ This enables advanced users to provide their own BF16 and FP32 operator list ins
     trainer.fit(model, datamodule=dm)
 
 For more details, please refer to `PyTorch Mixed Precision Training on Gaudi <https://docs.habana.ai/en/latest/PyTorch_User_Guide/PyTorch_User_Guide.html#pytorch-mixed-precision-training-on-gaudi>`__.
+
+----
+
+Enabling DeviceStatsMonitor
+---------------------------
```
Comment (kaushikb11) [inline docs/source-pytorch/accelerators/hpu_intermediate.rst:73]:
```suggestion
Enabling DeviceStatsMonitor with HPUs
----------------------------------------
```

File: src/pytorch_lightning/accelerators/hpu.py (hunk 2 above)
Comment (jerome-habana) [reply src/pytorch_lightning/accelerators/hpu.py:41]: added in the document

File: docs/source-pytorch/accelerators/hpu_intermediate.rst [hunk 4]
```diff
@@ -66,3 +66,34 @@ This enables advanced users to provide their own BF16 and FP32 operator list ins
...
 
 For more details, please refer to `PyTorch Mixed Precision Training on Gaudi <https://docs.habana.ai/en/latest/PyTorch_User_Guide/PyTorch_User_Guide.html#pytorch-mixed-precision-training-on-gaudi>`__.
+
+----
+
+Enabling DeviceStatsMonitor with HPUs
+----------------------------------------
+
+DeviceStatsMonitor is a callback that automatically monitors and logs device stats during training stage.
+This callback can be passed for training with HPUs. It returns a map of the following metrics with their values:
```
Comment (rohitgr7) [inline docs/source-pytorch/accelerators/hpu_intermediate.rst:76]:
```suggestion
:class:`~pytorch_lightning.callbacks.device_stats_monitor.DeviceStatsMonitor` is a callback that automatically monitors and logs device stats during the training stage.
This callback can be passed for training with HPUs. It returns a map of the following metrics with their values:
```

File: src/pytorch_lightning/CHANGELOG.md [hunk 5]
```diff
@@ -108,6 +108,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added support for DDP Fork ([#13405](https://github.com/PyTorchLightning/pytorch-lightning/pull/13405))
 
 
+- Added support for HPU Device stats monitor ([#13819](https://github.com/PyTorchLightning/pytorch-lightning/pull/13819))
```
Comment (rohitgr7) [inline src/pytorch_lightning/CHANGELOG.md:111]:
```suggestion
- Added support for HPU Device stats monitor ([#13819](https://github.com/Lightning-AI/lightning/pull/13819))
```

File: docs/source-pytorch/accelerators/hpu_intermediate.rst [hunk 6]
```diff
@@ -66,3 +66,34 @@ This enables advanced users to provide their own BF16 and FP32 operator list ins
...
+    **Limit** : amount of total memory on HPU device,
+    **InUse** : amount of allocated memory at any instance.,
+    **MaxInUse** : amount of total active memory allocated,
+    **NumAllocs** : number of allocations,
+    **NumFrees** : number of freed chunks,
+    **ActiveAllocs** : number of active allocations,
+    **MaxAllocSize** : maximum allocated size,
+    **TotalSystemAllocs** : total number of system allocations,
+    **TotalSystemFrees** : total number of system frees,
+    **TotalActiveAllocs** : total number of active allocations
```
Comment (rohitgr7) [inline docs/source-pytorch/accelerators/hpu_intermediate.rst:87]:
```suggestion
    **Limit** : amount of total memory on HPU device.
    **InUse** : amount of allocated memory at any instance.
    **MaxInUse** : amount of total active memory allocated.
    **NumAllocs** : number of allocations.
    **NumFrees** : number of freed chunks.
    **ActiveAllocs** : number of active allocations.
    **MaxAllocSize** : maximum allocated size.
    **TotalSystemAllocs** : total number of system allocations.
    **TotalSystemFrees** : total number of system frees.
    **TotalActiveAllocs** : total number of active allocations.
```

File: src/pytorch_lightning/accelerators/hpu.py [hunk 7]
```diff
@@ -39,9 +39,24 @@ def setup_environment(self, root_device: torch.device) -> None:
...
+        Limit,
+        InUse,
+        MaxInUse,
+        NumAllocs,
+        NumFrees,
+        ActiveAllocs,
+        MaxAllocSize,
+        TotalSystemAllocs,
+        TotalSystemFrees,
+        TotalActiveAllocs
```
Comment (rohitgr7) [inline src/pytorch_lightning/accelerators/hpu.py:53]: can you check the alignment in the docs here?

**Review by rohitgr7 — APPROVED** [review]

File: src/pytorch_lightning/accelerators/hpu.py [hunk 8]
```diff
@@ -39,9 +39,24 @@ def setup_environment(self, root_device: torch.device) -> None:
             raise MisconfigurationException(f"Device should be HPU, got {root_device} instead.")
 
     def get_device_stats(self, device: Union[str, torch.device]) -> Dict[str, Any]:
-        """HPU device stats aren't supported yet."""
-        rank_zero_debug("HPU device stats aren't supported yet.")
-        return {}
+        """Returns a map of the following metrics with their values:
+
+        Limit,
```
Comment (carmocca) [inline src/pytorch_lightning/accelerators/hpu.py:44]: Can this refer to the HPU docs instead of listing them? So it doesn't become outdated

File: tests/tests_pytorch/accelerators/test_hpu.py [hunk 9]
```diff
@@ -303,3 +303,23 @@ def training_epoch_end(self, outputs) -> None:
...
+        "NumAllocs",
+        "NumFrees",
+        "ActiveAllocs",
+        "MaxAllocSize",
+        "TotalSystemAllocs",
+        "TotalSystemFrees",
+        "TotalActiveAllocs",
+    ]
+    for f in fields:
+        assert any(f in h for h in hpu_stats.keys())
```
Comment (carmocca) [inline tests/tests_pytorch/accelerators/test_hpu.py:325]: I would test instead that it returns a list of strings and that some of the values exist (the basic ones). We don't want this to fail if the names are changed in torch_hpu

Comment (jerome-habana) [reply tests/tests_pytorch/accelerators/test_hpu.py:325]: As and when hpu package is upgraded, this will need change. Its better this way to catch the exceptions

File: src/pytorch_lightning/accelerators/hpu.py (hunk 8 above)
Comment (jerome-habana) [reply src/pytorch_lightning/accelerators/hpu.py:44]: Document link is provided in the doc. For the current version, this is what is supported. We dont have a design in place to upgrade the package separately.

File: src/pytorch_lightning/accelerators/hpu.py (hunk 7 above)
Comment (kaushikb11) [reply src/pytorch_lightning/accelerators/hpu.py:53]: @rohitgr7 what do you mean here?

**Review by kaushikb11 — APPROVED** [review]

File: src/pytorch_lightning/accelerators/hpu.py (hunk 7 above)
Comment (rohitgr7) [reply src/pytorch_lightning/accelerators/hpu.py]: updated

File: src/pytorch_lightning/accelerators/hpu.py [hunk 10]
```diff
@@ -39,9 +39,24 @@ def setup_environment(self, root_device: torch.device) -> None:
             raise MisconfigurationException(f"Device should be HPU, got {root_device} instead.")
 
     def get_device_stats(self, device: Union[str, torch.device]) -> Dict[str, Any]:
-        """HPU device stats aren't supported yet."""
-        rank_zero_debug("HPU device stats aren't supported yet.")
-        return {}
+        """Returns a map of the following metrics with their values:
+
+        - Limit
```
Comment (Borda) [inline src/pytorch_lightning/accelerators/hpu.py:44]: Can we also add types here like int or str?

**Review by Borda — APPROVED** [review]

File: src/pytorch_lightning/accelerators/hpu.py (hunk 10 above)
Comment (rohitgr7) [reply src/pytorch_lightning/accelerators/hpu.py:44]: I think all of them will be int/floats, else they can't be logged to loggers. We can add descriptions though.

File: docs/source-pytorch/accelerators/hpu_intermediate.rst [hunk 11]
```diff
@@ -66,3 +66,34 @@ This enables advanced users to provide their own BF16 and FP32 operator list ins
     trainer.fit(model, datamodule=dm)
 
 For more details, please refer to `PyTorch Mixed Precision Training on Gaudi <https://docs.habana.ai/en/latest/PyTorch_User_Guide/PyTorch_User_Guide.html#pytorch-mixed-precision-training-on-gaudi>`__.
+
+----
+
+Enabling DeviceStatsMonitor with HPUs
```
Comment (rohitgr7) [inline docs/source-pytorch/accelerators/hpu_intermediate.rst:72]: let's make sure this get's merged on master after the release.

Comment (kaushikb11) [reply docs/source-pytorch/accelerators/hpu_intermediate.rst:72]: @rohitgr7 What? We want it for the release.

Comment (rohitgr7) [reply docs/source-pytorch/accelerators/hpu_intermediate.rst:72]: @kaushikb11 
did you check with @carmocca !

I suggested this for v1.7 last week, but due to feature freeze it couldn't make it.

check the milestone.

Comment (kaushikb11) [reply docs/source-pytorch/accelerators/hpu_intermediate.rst:72]: @Borda What should we do about it? It's not a major feature but rather an enhancement to align the updates with the Habana API.

File: src/pytorch_lightning/accelerators/hpu.py (hunk 10 above)
Comment (jerome-habana) [reply src/pytorch_lightning/accelerators/hpu.py]: Its in bytes of type uint64

File: docs/source-pytorch/accelerators/hpu_intermediate.rst (hunk 11 above)
Comment (Borda) [reply docs/source-pytorch/accelerators/hpu_intermediate.rst:72]: If it is stable as it seems so, let's add it...
